src/include/catalog/catalog.h

```diff
@@ -0,0 +1,325 @@
+#pragma once
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include "catalog/catalog_defs.h"
+#include "catalog/catalog_sql_table.h"
+#include "catalog/database_handle.h"
+#include "catalog/tablespace_handle.h"
+#include "common/strong_typedef.h"
+#include "loggers/catalog_logger.h"
+#include "storage/sql_table.h"
+
+namespace terrier::catalog {
+
+class DatabaseHandle;
+class TablespaceHandle;
+class SettingsHandle;
+
+/**
+ * Schema column for used/unused schema rows.
+ * This is only used to transfer schema definition information to the sql table for Create. Thereafter, the
+ * this structure should not be used and schema information can be obtained from the catalog or from
+ * sql table GetSchema.
+ * Note that col_num is NOT used.
+ */
+struct SchemaCol {
+  /** column no */
+  int32_t col_num;
+  /** column name */
+  const char *col_name;
+  /** column type id */
+  type::TypeId type_id;
+};
+
+/**
+ * The global catalog object.
+ *
+ * The catalog is modeled upon the Postgres catalog, modified for terrier.
+ *
+ * pg_database, pg_tablespace and pg_settings are global catalogs
+ * (only a single instance). Other catalogs are "local" to each database.
+ * For example, a pg_class catalog exists for each database.
+ *
+ * OID assignment:
+ * In terrier, each table row is identified by an oid. Further, as strong
+ * typing is used in terrier, each (catalog) table has it's own oid type,
+ * e.g. db_oid_t, namespace_oid_t etc.
+ *
+ * The values for all oid types are generated from a single value space
+ * i.e. every oid regardless of type, is unique.
+ *
+ */
+class Catalog {
```

Comment:
It seems weird that one can use the `Catalog` object to create both databases and user tables. Why not namespaces? See my comments about `CreateUserTable()` and  `DeleteTable()` where you are not specifying a namespace, thereby violating the catalog hierarchy.

Reply:
Updated namespace support.